pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,149 +18,506 @@ import (
-func (t *Topology) Inject(ctx context.Context, constraints *v1alpha5.Constraints, pods []*v1.Pod) error {
-	// Group pods by equivalent topology spread constraints
-	topologyGroups := t.getTopologyGroups(pods)
-	// Compute spread
-	for _, topologyGroup := range topologyGroups {
-		if err := t.computeCurrentTopology(ctx, constraints, topologyGroup); err != nil {
-			return fmt.Errorf("computing topology, %w", err)
+type matchingTopology struct {
+	topology              *TopologyGroup
+	controlsPodScheduling bool
+	selectsPod            bool
+}
+
+// Initialize scans the pods provided and creates topology groups for any topologies that we need to track based off of
+// topology spreads, affinities, and anti-affinities specified in the pods.
+func (t *Topology) Initialize(ctx context.Context, pods ...*v1.Pod) error {
+	var errs error
+	errs = multierr.Append(errs, t.trackExistingAntiAffinities(ctx))
+	for _, p := range pods {
+		// need to first ensure that we know all of the topology constraints.  This may require looking up
+		// existing pods in the running cluster to determine zonal topology skew.
+		if err := t.trackTopologySpread(ctx, p); err != nil {
+			errs = multierr.Append(errs, fmt.Errorf("tracking topology spread, %w", err))
 		}
-		for _, pod := range topologyGroup.Pods {
-			domain := topologyGroup.NextDomain(constraints.Requirements.Add(v1alpha5.NewPodRequirements(pod).Requirements...).
-				Get(topologyGroup.Constraint.TopologyKey).
-				Values())
-			pod.Spec.NodeSelector = functional.UnionStringMaps(pod.Spec.NodeSelector, map[string]string{topologyGroup.Constraint.TopologyKey: domain})
+		if err := t.trackPodAffinityTopology(ctx, p); err != nil {
+			errs = multierr.Append(errs, fmt.Errorf("tracking affinity topology, %w", err))
+		}
+	}
+	return errs
+}
+
+// trackExistingAntiAffinities is used to identify pods with anti-affinity terms so we can track those topologies.  We
+// have to look at every pod in the cluster as there is no way to query for a pod with anti-affinity terms.
+func (t *Topology) trackExistingAntiAffinities(ctx context.Context) error {
+	// TODO: if can we determine somehow that the LimitPodHardAntiAffinityTopology admission controller is in place, we
+	// don't need to do any of this.  In that case anti-affinity would be hostname only which makes this tracking un-needed.
+	var nodeList v1.NodeList
+	if err := t.kubeClient.List(ctx, &nodeList); err != nil {
+		return fmt.Errorf("listing nodes, %w", err)
+	}
+
+	for _, node := range nodeList.Items {
+		var podlist v1.PodList
+		if err := t.kubeClient.List(ctx, &podlist, client.MatchingFields{"spec.nodeName": node.Name}); err != nil {
+			return fmt.Errorf("listing pods on node %s, %w", node.Name, err)
+		}
+		for _, p := range podlist.Items {
+			// lint warning is about taking the address of a variable in the for loop, it would matter if
+			// HasRequiredPodAntiAffinity modified p, but it doesn't
+			//nolint:gosec
```

Comment:
We typically avoid this pattern in general, since this will introduce a code bug if `HasRequiredPodAntiAffinity` starts modifying p. The way we typically avoid this is by indexing:

```
for i := range podlist.Items {
  if pod.HasRequiredPodAntiAffinity(&podlist.Items[i]) {
```